src/NuGet.Indexing/GalleryServiceImpl.cs

```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
             var searcher = searcherManager.Get();
             try
             {
-                bool includeUnlisted = ignoreFilter;
-                includePrerelease = ignoreFilter || includePrerelease;
+                // The old V2 search service would tread "id:" queries (~match) in the same way as it did "packageid:" (==match).
+                // If "id:" is in the query, replace it.
+                if (luceneQuery && !string.IsNullOrEmpty(q) && q.StartsWith("id:", StringComparison.OrdinalIgnoreCase))
+                {
+                    q = "packageid:" + q.Substring(3);
+                }
+
+                // The behavior below is incorrect when looking at the parameters for this method.
+                // One would expect the filters to be applied properly.
+                //
+                // We are *intentionally* doing it wrong here because our legacy V2 search service ignored
+                // these filters as well. The following URL's would yield the exact same result:
+                //   /search/query?q=Id:Antlr&prerelease=false&ignoreFilter=false
+                //   /search/query?q=Id:Antlr&prerelease=false&ignoreFilter=true
+                //   /search/query?q=Id:Antlr&prerelease=true&ignoreFilter=false
+                //   /search/query?q=Id:Antlr&prerelease=true&ignoreFilter=true
+                //
+                // If this behavior needs to be changed, this is the bit of code that will return
+                // the data you'd expect looking at the method parameters:
+                //   bool includeUnlisted = ignoreFilter;
+                //   includePrerelease = ignoreFilter || includePrerelease;
+                //   feed = ignoreFilter ? null : feed;
+                //   Filter filter = searcher.GetFilter(includeUnlisted, includePrerelease, feed);
+                //
+                // But hey, we'll do it the old V2 way instead and happily ignore the correct way.
                 feed = ignoreFilter ? null : feed;
-                Filter filter = searcher.GetFilter(includeUnlisted, includePrerelease, feed);
+                Filter filter = searcher.GetFilter(feed);
 
-                Query query = NuGetQuery.MakeQuery(q);
+                Query query = NuGetQuery.MakeQuery(q, searcher);
+
+                if (filter != null)
+                {
+                    query = new FilteredQuery(query, filter);
+                    filter = null;
```

Comment:
This filters before running the query, reducing the search set (and the amount of results to search a score/ranking for). Adding a comment.